Us 11 13 - Inventory System

Alexander-LF opened the pull request.

Alexander-LF committed:
Added US-13

Inventory system implementation from US-13, except for the UI.

Alexander-LF committed:
Minor fixes to inventory

Alexander-LF committed:
Merge branch 'US-11-13' into US-13

Alexander-LF committed:
Merge pull request #5 from james-love/US-13

Us 13

Alexander-LF committed:
Added Inventory UI

Added a UI for the inventory, along with some inventory object implementations.

Alexander-LF committed:
Fixes for inventory UI

Fixed issue where secondary thread that updated inventory UI could crash the game when trying to open a second inventory form. Added a method to the IWeapon interface.

james-love commented on Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj:
```diff
@@ -8,10 +8,11 @@
     <RootNamespace>Fall2020_CSC403_Project</RootNamespace>
     <AssemblyName>Fall2020_CSC403_Project</AssemblyName>
-    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
What's the reason for downgrading the framework?

james-love commented on Project/Fall2020_CSC403_Project/Fall2020_CSC403_Project.csproj:
```diff
@@ -60,14 +67,17 @@
     </EmbeddedResource>
     <EmbeddedResource Include="Properties\Resources.resx">
-      <Generator>ResXFileCodeGenerator</Generator>
+      <Generator>PublicResXFileCodeGenerator</Generator>
```
is this related to this issue?
https://stackoverflow.com/questions/4274311/visual-studio-resx-file-default-internal-to-public

Alexander-LF replied:
Forgot to roll back the changes to this when from when I tried accessing the inventory images from the file instead of the form.

james-love commented on Project/Fall2020_CSC403_Project/FrmLevel.cs:
```diff
@@ -147,6 +156,17 @@ private void OnEnemyDeath(Enemy enemy)
+            break;
+
+        case Keys.C:
+          player.GetInventory().AddToInventory(new MyGameLibrary.InventoryObjects.HealthPotion(picHealthPotion.Image));
```
Instead of homemade images, consider using free to use artwork such as this https://opengameart.org/content/16-bit-rpg-potion-pack

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+        {
+            inventory = new InventoryObject[maxInventorySize];
+            this.maxInventorySize = maxInventorySize;
+            for (int i = 0; i < invObjs.Count && i < maxInventorySize; i++)
```
Using a foreach would look cleaner. This is true of the other constructor as well.

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+
+        public int GetMaxInventorySize()
+        {
+            return maxInventorySize;
```
Instead of adding public accessor methods for private variable, consider something like this

```
private int _maxSize = 8;
public int MaxSize { get => _maxSize; }
```

Also, since its inside the inventory class, the full name of MaxInventorySize could be shortened to just MaxSize.

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Several of this using statements aren't actually used in this file and should be removed. This is also true of the other new files in this review, so those should be checked as well. In Visual Studio you can quickly tell which ones are not being used as they will be slightly translucent and when you hover over them with the mouse it will tell you that they are unused.

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+
+        // Returns the index of the given object type (matches the passed
+        // object's class) in the inventory.
+        private int ObjectTypeExistsInInventory(InventoryObject iObj)
```
The name indicates that this would be a boolean check, however from the code this is really a find index returning -1 when there is none found. Consider replacing the function with this one that accomplishes the same task:
`return Array.FindIndex<InventoryObject>(inventory, x => x != null && x.TypeOf(iObj));`

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+        // Returns object in inventory at index without removing it.
+        public InventoryObject GetInventoryObject(int index)
+        {
+            //Console.WriteLine("\t" + index.ToString());
```
Remove unused commented out code

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+            else
+            {
+                return null;
+            }
```
Its a matter of preference here on if you want to change it but this could be like this instead
```
if (index < inventory.Count())
    return (inventory[index]);
else
    return null;
```

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+        public InventoryObject GetInventoryObject(int index)
+        {
+            //Console.WriteLine("\t" + index.ToString());
+            if (index < inventory.Count())
```
This handles if the user passes and index larger than the array, but what if they pass a negative index? Consider instead wrapping this in a try catch and returning null on exception, taking care of this and any other case of an invalid index being passed

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+        // Removes an object at index from the inventory
+        public void RemoveInventoryObject(int index)
+        {
+            if(index < maxInventorySize)
```
Similar to the other comment, this doesn't handle negative numbers. See other comment for proposed solution.

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+        // Removes the given object from the inventory
+        public void RemoveInventoryObject(InventoryObject iObj)
+        {
+            for(int i = 0; i < maxInventorySize; i++)
```
Consider using a foreach instead

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+        public void UseInventoryObject(int index, Player player)
+        {
+            InventoryObject iObj = GetInventoryObject(index);
+            if(iObj == null)
```
In general, there are several places with if statements that could have the braces removed, look over all the new code and determine which ones to remove.

Alexander-LF replied:
I'm going to keep with the Google C# style guide we were recommended, which specifies to always use braces when optional.

james-love commented on Project/MyGameLibrary/Inventory.cs:
```diff
@@ -0,0 +1,218 @@
+                return;
+            }
+            iObj.Effect(player);
+            if (iObj.IsExhaustible())
```
All these if statements could be shortened to
```
if (iObj.IsExhaustible() && (!iObj.IsStackable() || iObj.GetCount() <= 0)
    RemoveInventoryObject(index);
```

james-love commented on Project/MyGameLibrary/InventoryItems/Experience.cs:
```diff
@@ -0,0 +1,34 @@
+            this.exhaustible = true;
+            this.img = img;
+            this.count = count;
+            
```
Extra new line added here

james-love commented on Project/MyGameLibrary/InventoryItems/HealthPotion.cs:
```diff
@@ -0,0 +1,34 @@
+            this.stackable = false;
+            this.exhaustible = true;
+            this.img = img;
+            
```
Extra new line added here

james-love commented on Project/MyGameLibrary/InventoryItems/HealthPotion.cs:
```diff
@@ -0,0 +1,34 @@
+        public override void Effect(BattleCharacter character)
+        {
+            throw new NotImplementedException();
+
```
Extra new line added here, there are several places in other files as well. After making the other changes from this review, run through the code and remove them

james-love commented on Project/MyGameLibrary/InventoryObject.cs:
```diff
@@ -0,0 +1,87 @@
+        }
+
+        // Returns if this object is stackable
+        public bool IsStackable()
```
Instead of creating methods like this to access these fields, it would make more sense for them to be treated as properties for the object. This statement applies to all of the IsWhatever variables in this class as well as to the get count. Consider this
```
protected bool _stackable; 
public bool Stackable { get => stackable; }
```
Now later on instead of addressing the values as 
```
iObj.IsStackable()
iObj.IsExhaustible()
iObj.GetCount()
```
you can instead just use
```
iObj.Stackable
iObj.Exhaustible
iObj.Count
```

james-love commented on Project/MyGameLibrary/MyGameLibrary.csproj:
```diff
@@ -9,9 +9,10 @@
     <RootNamespace>MyGameLibrary</RootNamespace>
     <AssemblyName>MyGameLibrary</AssemblyName>
-    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
Why has the framework version been downgraded?

Alexander-LF replied:
My mistake, this is from using an older version of VS on my machine. It shouldn't change any functionality, but I'll update this before I push my updated code.

james-love commented on Project/MyGameLibrary/Player.cs:
```diff
@@ -1,13 +1,22 @@
+            inventory = new Inventory();
+    }
 
+    public Inventory GetInventory()
```
Similar to the comment I made in the InventoryObject class, should just use a property here instead of a separate accessor function.

Alexander-LF replied:
This would cause a conflict where the class name is the same as the property.

james-love commented on Project/Fall2020_CSC403_Project/FrmInventory.cs:
```diff
@@ -0,0 +1,167 @@
+
+namespace Fall2020_CSC403_Project
+{
+    public partial class FrmInventory : Form
```
The inventory window currently shows up with a title of "Form1" this can be changed in the designer by selecting the overall window/form and changing the "Text" property.

james-love commented on Project/Fall2020_CSC403_Project/FrmInventory.cs:
```diff
@@ -0,0 +1,167 @@
+            inventorySlots = new Button[inventory.GetMaxInventorySize()];
+            InitInvSlots();
+            this.FormClosing += new FormClosingEventHandler(FrmIventory_FormClosing);
+            invUpdater = new Thread(ContinuousUpdater);
```
There is a bug which I think may be caused by this  where if you open the inventory then close the main game window without closing inventory first, you can no longer rebuild the project as the files will be locked by the still running process in the background which has to be manually killed.

james-love commented on Project/Fall2020_CSC403_Project/FrmInventory.cs:
```diff
@@ -0,0 +1,167 @@
+                                inventorySlots[i].Text = inventory.GetInventoryObject(i).GetCount().ToString();
+                            }));
+                        }
+                        catch(Exception e)
```
Should we be doing anything when we catch all these exceptions? If we're not specifying the exception on any of these or doing anything once it's caught, why not wrap the whole thing in a single try catch?

Alexander-LF committed:
Fixed issues with formatting and thread safety in inventory

Formatting fixes accross the inventory update, and removed the secondary inventory UI updating thread.

Alexander-LF committed:
Fixed framework version and resources file accessibility

Alexander-LF committed:
Merge branch 'main' into US-11-13

Alexander-LF committed:
Bug and style fixes

Changed project resx file to internal, changed some variables to properties, and fixed and issue where the inventory wasn't initializing before being manually opened.

Alexander-LF committed:
Merge branch 'US-11-13' of https://github.com/james-love/Fall2020_CSC403_Project into US-11-13

james-love reviewed (approved):
looks good
